C-DataStructures/ArrayStack/ArrayStack.c

```diff
@@ -0,0 +1,245 @@
+/**
+ * @file ArrayStack.c
+ *
+ * @brief Implements the functions of the ArrayStack Structure.
+ *
+ * @ingroup C-DataStructure
+ *
+ * @author Edward Lon
+ * Contact: edwardlon@outlook.com
+ *
+ */
+
+#include "ArrayStack.h"
+
+/* See header for documentation. */
+bool arrayStackInit( ArrayStack** as )
+{
+    bool status = false;
+
+    if (*as != NULL)
+    {
+        fprintf(stderr, "Already initialized!\n\r");
+    }
+    else if ((*as = (ArrayStack*)malloc(sizeof(ArrayStack) * C_MAX_STACK_SIZE)) == NULL)
+    {
+        fprintf(stderr, "Failed to malloc memory!");
+    }
+    else if (arrayStackInitArray(*as) == false)
+    {
+        fprintf(stderr, "Failed to malloc array!\n\r");
+    }
+    else
+    {
+        (*as)->m_size = 0;
+
+        status = true;
+    }
+
+    return status;
+}
+
+/* See header for documentation. */
+bool arrayStackInitArray( ArrayStack* as )
+{
+    bool status = false;
+
+    if (as == NULL)
+    {
+        fprintf(stderr, "ArrayStack is null!\n\r");
+    }
+    else if ((as->m_array = (int32_t**)malloc(sizeof(int32_t*) * C_MAX_STACK_SIZE)) == NULL)
+    {
+        fprintf(stderr, "Failed to malloc space for the array!\n\r");
+    }
+    else
+    {
+        for (uint32_t i = 0; i < C_MAX_STACK_SIZE; i++)
+        {
+            as->m_array[i] = (int32_t*)malloc(sizeof(int32_t));
+
+            if (as->m_array[i] == NULL)
+            {
+                fprintf(stderr, "Out of memory!\n\r");
+                status = false;
+                break;
+            }
+        }
+        status = true;
+    }
+
+    return status;
+}
+
+/* See header for documentation. */
+bool arrayStackPush( ArrayStack* as, const int32_t value )
+{
+    bool status = false;
+
+    if (as == NULL)
+    {
+        fprintf(stderr, "ArrayStack is not initialized!\n\r");
+    }
+    else if (as->m_size + 2 > C_MAX_STACK_SIZE)
```

Comment:
Any reason for +2 size?

Reply:
2 is to account for the array index and the pre-increment down below so that we don't write to some region of space